From 51f590b5f7bb15439d2baf70a4c3778f063966f6 Mon Sep 17 00:00:00 2001
From: Rahul Akolkar <rahul@apache.org>
Date: Fri, 14 Nov 2008 23:43:16 +0000
Subject: [PATCH] <assign> doesn't remove all children of old node. A variant
 of a patch thanks to Joel Truher <joel at truher dot org>. Adding Joel to
 list of contributors. Also fixed a minor formatting inconsistency in a
 related error message. SCXML-89

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/scxml/trunk@714191 13f79535-47bb-0310-9956-ffa450edef68
---
 pom.xml                                       |  3 ++
 project.xml                                   |  3 ++
 .../org/apache/commons/scxml/Builtin.java     |  4 +-
 .../apache/commons/scxml/model/Assign.java    |  9 ++--
 .../commons/scxml/model/AssignTest.java       | 21 +++++++--
 .../{assign-test.xml => assign-test-01.xml}   |  0
 .../commons/scxml/model/assign-test-02.xml    | 46 +++++++++++++++++++
 7 files changed, 76 insertions(+), 10 deletions(-)
 rename src/test/java/org/apache/commons/scxml/model/{assign-test.xml => assign-test-01.xml} (100%)
 create mode 100644 src/test/java/org/apache/commons/scxml/model/assign-test-02.xml

diff --git a/pom.xml b/pom.xml
index 6b2111894..b885dcd17 100644
--- a/pom.xml
+++ b/pom.xml
@@ -117,6 +117,9 @@
     <contributor>
       <name>Jakob Sachse</name>
     </contributor>
+    <contributor>
+      <name>Joel Truher</name>
+    </contributor>
   </contributors>
 
   <dependencies>
diff --git a/project.xml b/project.xml
index 3ca19451f..689d2a177 100644
--- a/project.xml
+++ b/project.xml
@@ -164,6 +164,9 @@
     <contributor>
       <name>Jakob Sachse</name>
     </contributor>
+    <contributor>
+      <name>Joel Truher</name>
+    </contributor>
   </contributors>
   
   <dependencies>
diff --git a/src/main/java/org/apache/commons/scxml/Builtin.java b/src/main/java/org/apache/commons/scxml/Builtin.java
index c440121b3..96846ec67 100644
--- a/src/main/java/org/apache/commons/scxml/Builtin.java
+++ b/src/main/java/org/apache/commons/scxml/Builtin.java
@@ -126,7 +126,7 @@ public static Node dataNode(final Map namespaces, final Object data,
         } else {
             if (length > 1) {
                 Log log = LogFactory.getLog(Builtin.class);
-                log.warn("Data(): Multiple nodes matching XPath expression"
+                log.warn("Data(): Multiple nodes matching XPath expression \""
                     + path + "\", returning first");
             }
             return result.item(0);
@@ -209,7 +209,7 @@ public static Node dataNode(final Object data, final String path) {
         } else {
             if (length > 1) {
                 Log log = LogFactory.getLog(Builtin.class);
-                log.warn("Data(): Multiple nodes matching XPath expression"
+                log.warn("Data(): Multiple nodes matching XPath expression \""
                     + path + "\", returning first");
             }
             return result.item(0);
diff --git a/src/main/java/org/apache/commons/scxml/model/Assign.java b/src/main/java/org/apache/commons/scxml/model/Assign.java
index 8c3a48273..fb7f8359e 100644
--- a/src/main/java/org/apache/commons/scxml/model/Assign.java
+++ b/src/main/java/org/apache/commons/scxml/model/Assign.java
@@ -196,10 +196,11 @@ public void execute(final EventDispatcher evtDispatcher,
                         newNode = eval.evalLocation(ctx, expr);
                     }
                     // Remove all children
-                    for (Node child = oldNode.getFirstChild();
-                            child != null;
-                            child = child.getNextSibling()) {
-                        oldNode.removeChild(child);
+                    Node removeChild = oldNode.getFirstChild();
+                    while (removeChild != null) {
+                        Node nextChild = removeChild.getNextSibling();
+                        oldNode.removeChild(removeChild);
+                        removeChild = nextChild;
                     }
                     if (newNode != null) {
                         // Adopt new children
diff --git a/src/test/java/org/apache/commons/scxml/model/AssignTest.java b/src/test/java/org/apache/commons/scxml/model/AssignTest.java
index 261d67b7f..7abeb9303 100644
--- a/src/test/java/org/apache/commons/scxml/model/AssignTest.java
+++ b/src/test/java/org/apache/commons/scxml/model/AssignTest.java
@@ -45,21 +45,24 @@ public static Test suite() {
     }
 
     // Test data
+    private URL assign01, assign02;
     private SCXMLExecutor exec;
 
     /**
      * Set up instance variables required by this test case.
      */
     public void setUp() {
-        URL assignSample = this.getClass().getClassLoader().
-            getResource("org/apache/commons/scxml/model/assign-test.xml");
-        exec = SCXMLTestHelper.getExecutor(assignSample);
+        assign01 = this.getClass().getClassLoader().
+            getResource("org/apache/commons/scxml/model/assign-test-01.xml");
+        assign02 = this.getClass().getClassLoader().
+            getResource("org/apache/commons/scxml/model/assign-test-02.xml");
     }
 
     /**
      * Tear down instance variables required by this test case.
      */
     public void tearDown() {
+        assign01 = assign02 = null;
         exec = null;
     }
 
@@ -67,6 +70,7 @@ public void tearDown() {
      * Test the implementation
      */
     public void testAssignSrc() {
+        exec = SCXMLTestHelper.getExecutor(assign01);
         Set currentStates = exec.getCurrentStatus().getStates();
         assertEquals(1, currentStates.size());
         assertEquals("assign3", ((State)currentStates.iterator().
@@ -74,7 +78,16 @@ public void testAssignSrc() {
         assertTrue(exec.getCurrentStatus().isFinal());
     }
 
-     public static void main(String args[]) {
+    public void testAssignDeep() {
+        exec = SCXMLTestHelper.getExecutor(assign02);
+        Set currentStates = exec.getCurrentStatus().getStates();
+        assertEquals(1, currentStates.size());
+        assertEquals("assign3", ((State)currentStates.iterator().
+            next()).getId());
+        assertTrue(exec.getCurrentStatus().isFinal());
+    }
+
+    public static void main(String args[]) {
         TestRunner.run(suite());
     }
 }
diff --git a/src/test/java/org/apache/commons/scxml/model/assign-test.xml b/src/test/java/org/apache/commons/scxml/model/assign-test-01.xml
similarity index 100%
rename from src/test/java/org/apache/commons/scxml/model/assign-test.xml
rename to src/test/java/org/apache/commons/scxml/model/assign-test-01.xml
diff --git a/src/test/java/org/apache/commons/scxml/model/assign-test-02.xml b/src/test/java/org/apache/commons/scxml/model/assign-test-02.xml
new file mode 100644
index 000000000..38ab9b3dc
--- /dev/null
+++ b/src/test/java/org/apache/commons/scxml/model/assign-test-02.xml
@@ -0,0 +1,46 @@
+<?xml version="1.0"?>
+<!--
+   * Licensed to the Apache Software Foundation (ASF) under one or more
+   * contributor license agreements.  See the NOTICE file distributed with
+   * this work for additional information regarding copyright ownership.
+   * The ASF licenses this file to You under the Apache License, Version 2.0
+   * (the "License"); you may not use this file except in compliance with
+   * the License.  You may obtain a copy of the License at
+   *
+   *     http://www.apache.org/licenses/LICENSE-2.0
+   *
+   * Unless required by applicable law or agreed to in writing, software
+   * distributed under the License is distributed on an "AS IS" BASIS,
+   * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   * See the License for the specific language governing permissions and
+   * limitations under the License.
+-->
+<!-- Regress JIRA 89, incomplete child removal -->
+<scxml xmlns="http://www.w3.org/2005/07/scxml" xmlns:rad="http://foo/bar" version="1.0" initialstate="assign1">
+   <datamodel>
+       <data name="source">
+           <rad:foo>
+               <rad:a>1</rad:a>
+               <rad:b>2</rad:b>
+           </rad:foo>
+       </data>
+       <data name="destination">
+           <rad:bar>
+               <rad:a>3</rad:a>
+               <rad:b>4</rad:b>
+           </rad:bar>
+       </data>
+   </datamodel>
+   <!-- verify the destination contents -->
+   <state id="assign1" final="true">
+       <transition cond="Data(destination,'rad:bar/rad:a') eq 3 and Data(destination,'rad:bar/rad:b') eq 4" target="assign2" />
+   </state>
+   <!-- copy the new contents and verify -->
+   <state id="assign2">
+       <onentry>
+           <assign location="Data(destination,'rad:bar')" expr="Data(source,'rad:foo')" />
+       </onentry>
+       <transition cond="Data(destination,'rad:bar/rad:a') eq 1 and Data(destination,'rad:bar/rad:b') eq 2" target="assign3" />
+   </state>
+   <state id="assign3" final="true" />
+</scxml>
